lib/remote_console/rack_server.rb

```diff
@@ -0,0 +1,118 @@
+# This server allows the user to access remote endpoints through an HTTP(S)
+# connection to the appliance. Opposed to a direct access or a forwarded port
+# it is not necessary to expose an extra IP:port combination to the client.
+#
+# It uses socket hijacking to retrieve the underlying TCP socket connection
+# of the HTTP request incoming from a client. This socket is detached from
+# the Rack middleware and handled by a separate transmitter thread.
+# The remote address of the console is determined by a one time secret that
+# is part of the URL and it points to a SystemConsole record in the database.
+#
+# On the lowest level the transmitter thread operates with regular sockets,
+# however, both on the client and server side wrappers are used to translate
+# between a pair of sockets. Due to some limitations in the websocket driver
+# gem, the reading/writing operations are defined in the fetch/issue methods.
+#
+# The selector for the sockets is provided by an external gem and it handles
+# the dependency between the socket pairs, i.e. one should be ready to read
+# while other one should be ready to write at the same time. The pairs ready
+# for transmission are handled by the `each_ready` iterator. As the iterator
+# always returns with a socket to read and a socket to write, the `@adapters`
+# hash has been used to access the corresponding wrappers.
+
+module RemoteConsole
+  class RackServer
+    attr_accessor :logger
+
+    RACK_404 = [404, {'Content-Type' => 'text/plain'}, ['Not found']].freeze
+    RACK_YAY = [-1, {}, []].freeze
+
+    def initialize(options = {})
+      @logger = options.fetch(:logger, $remote_console_log)
+      @logger.info('Initializing websocket server!')
+      @proxy = SurroGate.new(logger)
+      @adapters = {}
+
+      @transmitter = Thread.new do
+        loop do
+          @proxy.select(1000)
+
+          @proxy.each_ready do |left, right|
+            begin
+              @adapters[left].fetch(64.kilobytes) { |data| @adapters[right].issue(data) } # left -> right
+            rescue IOError, IO::WaitReadable, IO::WaitWritable
+              cleanup(:info, "Closing websocket proxy for VM %{vm_id}", left, right)
+            rescue StandardError => ex
+              cleanup(:error, "Websocket proxy for VM %{vm_id} errored with #{ex} #{ex.backtrace.join("\n")}", left, right)
+            end
+          end
+        end
+      end
+
+      @transmitter.abort_on_exception = true
+    end
+
+    def call(env)
+      exp = %r{^/ws/console/([a-zA-Z0-9]+)/?$}.match(env['REQUEST_URI'])
+      if WebSocket::Driver.websocket?(env) && same_origin_as_host?(env) && exp.present?
+        @logger.info("Remote console connection initiated with secret #{exp[1]}")
```

Comment:
Should we be logging the secret?

Reply:
Well, good question ... it is dumb one-time password, not that much as a secret. Even if we don't want to log it, this is PR just moving code and I'd be more comfortable with doing it in a followup PR.